File: 2011_final_value_after_operations/final_value_after_operations.go
/*
 * @Date: 2022-12-23
 * @LastEditors: mengsen_wang@163.com
 * @LastEditTime: 2022-12-23
 * @FilePath: /algorithm/2011_final_value_after_operations/final_value_after_operations.go
 */

package main

func finalValueAfterOperations(operations []string) (x int) {
	for _, op := range operations {
		if op[1] == '+' {
			x++
		} else {
			x--
		}
	}
	return
}

func main() {
	assert := func(b bool) {
		if !b {
			panic("Not Passed")
		}
	}
	{
		operations := []string{"--X", "X++", "X++"}
		ans := 1
		assert(finalValueAfterOperations(operations) == ans)
	}

	{
		operations := []string{"++X", "++X", "X++"}
		ans := 3
		assert(finalValueAfterOperations(operations) == ans)
	}

	{
		operations := []string{"X++", "++X", "--X", "X--"}
		ans := 0
		assert(finalValueAfterOperations(operations) == ans)
	}
}

File: rust/src/2011_final_value_after_operations/final_value_after_operations.rs
/*
 * @Date: 2022-12-23
 * @LastEditors: mengsen_wang@163.com
 * @LastEditTime: 2022-12-23
 * @FilePath: /algorithm/2011_final_value_after_operations/final_value_after_operations.rs
 */

pub fn final_value_after_operations(operations: Vec<String>) -> i32 {
    operations.iter().fold(0, |mut x: i32, op: &String| {
        match op.chars().nth(1).unwrap() {
            '-' => x -= 1,
            _ => x += 1,
        }
        x
    })
}

fn main() {
    {
        let operations = vec!["--X", "X++", "X++"]
            .iter()
            .map(|s| s.to_string())
            .collect();
        let ans = 1;
        assert_eq!(final_value_after_operations(operations), ans);
    }

    {
        let operations = vec!["++X", "++X", "X++"]
            .iter()
            .map(|s| s.to_string())
            .collect();
        let ans = 3;
        assert_eq!(final_value_after_operations(operations), ans);
    }

    {
        let operations = vec!["X++", "++X", "--X", "X--"]
            .iter()
            .map(|s| s.to_string())
            .collect();
        let ans = 0;
        assert_eq!(final_value_after_operations(operations), ans);
    }
}
